cmd/frontend/internal/inventory/entries.go

```diff
@@ -61,39 +74,67 @@ func (c *Context) tree(ctx context.Context, tree fs.FileInfo, buf []byte) (inv I
 	}
 
 	entries, err := c.ReadTree(ctx, tree.Name())
+	sem.Release(1)
 	if err != nil {
 		return Inventory{}, err
 	}
 	invs := make([]Inventory, len(entries))
+	results := make(chan treeIteratorResult, len(entries)) // Buffer the channel to the number of entries
+	var wg sync.WaitGroup
+
 	for i, e := range entries {
-		switch {
-		case e.Mode().IsRegular(): // file
-			// Don't individually cache files that we found during tree traversal. The hit rate for
-			// those cache entries is likely to be much lower than cache entries for files whose
-			// inventory was directly requested.
-			lang, err := getLang(ctx, e, buf, c.NewFileReader)
-			if err != nil {
-				return Inventory{}, errors.Wrapf(err, "inventory file %q", e.Name())
-			}
-			invs[i] = Inventory{Languages: []Lang{lang}}
+		wg.Add(1)
+		go func(i int, e os.FileInfo) {
+			defer wg.Done()
 
-		case e.Mode().IsDir(): // subtree
-			subtreeInv, err := c.tree(ctx, e, buf)
-			if err != nil {
-				return Inventory{}, errors.Wrapf(err, "inventory tree %q", e.Name())
+			switch {
+			case e.Mode().IsRegular(): // file
+				// Don't individually cache files that we found during tree traversal. The hit rate for
+				// those cache entries is likely to be much lower than cache entries for files whose
+				// inventory was directly requested.
+				if err := sem.Acquire(ctx, 1); err != nil {
+					results <- treeIteratorResult{i, Inventory{}, err}
+				}
+				lang, err := getLang(ctx, e, buf, c.NewFileReader)
+				sem.Release(1)
+				if err != nil {
+					results <- treeIteratorResult{i, Inventory{Languages: []Lang{lang}}, err}
+				}
+				results <- treeIteratorResult{i, Inventory{Languages: []Lang{lang}}, nil}
+
+			case e.Mode().IsDir(): // subtree
+				subtreeInv, err := c.tree(ctx, e, buf, sem)
+				if err != nil {
+					results <- treeIteratorResult{i, subtreeInv, err}
+				}
+				results <- treeIteratorResult{i, subtreeInv, nil}
```

Comment:
No need to check `err` since we just pass the result along anyways.

```suggestion
				results <- treeIteratorResult{i, subtreeInv, err}
```